plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeTransactionLogEntryIterator.java

```diff
@@ -0,0 +1,121 @@
+    private Optional<Iterator<DeltaLakeTransactionLogEntry>> currentEntries = Optional.empty();
+    private final long startVersion;
+    private final Optional<Long> endVersion;
+
+    private final Optional<Long> lastCheckpointVersion;
+
+    public DeltaLakeTransactionLogEntryIterator(
+            TrinoFileSystem fileSystem,
+            Path tableLocation,
+            Optional<Long> startVersion,
+            Optional<Long> endVersion)
+    {
+        verify(startVersion.orElse(0L) <= endVersion.orElse(startVersion.orElse(0L)), "startVersion must be less or equal than endVersion");
+        currentVersion = startVersion.orElse(0L) - 1;
+        this.startVersion = startVersion.orElse(0L);
+        this.endVersion = endVersion;
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        transactionLogDir = getTransactionLogDir(requireNonNull(tableLocation, "tableLocation is null"));
+        lastCheckpointVersion = readLastCheckpoint(fileSystem, tableLocation).map(LastCheckpoint::getVersion);
+    }
+
+    @Override
+    public boolean hasNext()
+    {
+        if (currentEntries.isPresent()) {
+            if (currentEntries.get().hasNext()) {
+                return true;
+            }
+            else {
+                if (endVersion.isPresent() && (currentVersion == endVersion.get())) {
+                    return false;
+                }
+            }
+        }
+        if (currentVersion >= startVersion &&
+                currentEntries.isEmpty() &&
+                (lastCheckpointVersion.isEmpty() || currentVersion >= lastCheckpointVersion.get())) {
+            // Do not retrieve the next transaction log entry when the current one is missing
+            return false;
+        }
+
+        while (true) {
+            // Retrieve next transaction log entry
+            currentVersion++;
+            try {
+                currentEntries = getEntriesFromJson(currentVersion, transactionLogDir, fileSystem)
+                        .map(List::iterator);
+            }
+            catch (IOException e) {
+                throw new TrinoException(DELTA_LAKE_INVALID_SCHEMA, "Failed accessing transaction log " + getTransactionLogJsonEntryPath(transactionLogDir, currentVersion), e);
+            }
+
+            if (currentEntries.isPresent()) {
+                return true;
+            }
+            // The outdated transaction log file may have been removed after adding a subsequent checkpoint
+            if (lastCheckpointVersion.isPresent() && currentVersion < lastCheckpointVersion.get()) {
+                if (endVersion.isPresent() && currentVersion == endVersion.get()) {
+                    return false;
+                }
+                continue;
+            }
+            if (endVersion.isPresent()) {
+                return false;
+            }
+            return false;
+        }
+    }
+
+    @Override
+    public DeltaLakeTransactionLogEntry next()
+    {
+        if (hasNext()) {
```

Comment:
Is it necessary to verify the presence of the next element using `hasNext` at this point? Shouldn't it be the responsibility of the caller to invoke `hasNext()` prior to invoking `next()`?

Reply:
Modified the logic to avoid calling `hasNext()` in `next()`.